### DISCUSSION TITLE
Predefined Topics

### DISCUSSION THREAD

**Discussion by theobouwman** [MaartenGr/BERTopic#2167] [Q&A]
We observe that Events created in our platform can be classified as Career, Tech, Cultural or Talks when we look at the created topics.

Is it possible to predefine these topics, and let BERTopic classify all documents, existing and new, to these predefined topics?

**Comment by MaartenGr** [comment]
Yes, you can! If you use [zero-shot topic modeling](https://maartengr.github.io/BERTopic/getting_started/zeroshot/zeroshot.html), you can specify these topics and then set the minimum similarity to 0. That will in turn classify all documents with those predefined topics.

**Comment by theobouwman** [comment]
How specific should these topics be defined. Something like this:
```
zeroshot_topic_list = [
    "Career opportunities job employer student", 
    "Digital ICT Technology AI Robot",
    "Politics International Relations Talks Lectures Professor University",
    "Culture Arts Performance Geographer Dance Theater Festival Performances",
]
```

Or use topcis like "Career Jobs", "Digitial ICT AI", "Politics Geo...", "Cultural"?

Also with this code and the latest bertopic version:

```python
umap_model = UMAP(n_neighbors=15, n_components=5, min_dist=0.0, metric='cosine', random_state=42)


topic_model = BERTopic(
    embedding_model="thenlper/gte-small", 
    zeroshot_topic_list=zeroshot_topic_list,
    zeroshot_min_similarity=0,
    representation_model=KeyBERTInspired(),
    umap_model=umap_model # reproducability
)

topics, _ = topic_model.fit_transform(docs_en)

freq = topic_model.get_topic_info()
```

we get:

<img width="1161" alt="Screenshot 2024-10-23 at 20 22 30" src="https://github.com/user-attachments/assets/a097b19c-0e60-4bd4-b661-69489c085a69">

The Name column is not set to the topic, and the order is mixed up.

And last question. The Representation column, are that the most common words per topic? If so, is it possible to get this on a per document basis? I am asking this because I have also experimented. with KeyBERT to get keyworks per document, but maybe it is more convenient to use the BERTopic library for this as well.

Thanks

**Reply by MaartenGr** [reply]
> Or use topcis like "Career Jobs", "Digitial ICT AI", "Politics Geo...", "Cultural"?

Either work but note that these texts are embedded, so providing additional detail will help but might make the resulting embedding more specific as a result. It's up to you to decide how abstract/specific you would want a topic to be. 

> Also with this code and the latest bertopic version:

Which BERTopic version are you using exactly? There was a fix in v0.16.4 and I need to be sure you are using that one.

> And last question. The Representation column, are that the most common words per topic? If so, is it possible to get this on a per document basis? I am asking this because I have also experimented. with KeyBERT to get keyworks per document, but maybe it is more convenient to use the BERTopic library for this as well.

Those are not the most common words per topic but the words extracted with `KeyBERTInspired`. To do this on a per-document level, you would indeed have to use the separate KeyBERT package since BERTopic focus on groups of documents rather than individual documents.

**Reply by alihashaam** [reply]
Hey @MaartenGr 

First thanks for the great work on one side and also for help and feedback here.

I am facing the same issue of topics from `zeroshot_topic_list ` not appearing in `Name` column, after running `get_topic_info`. And I am not sure if I need to parse labels in some way before?

First I thought of running zeroshotclassification with representation_models but then saw in some other threads you have suggested not to use that after doing Zeroshot modelling. Feedback will be highly appreciated

Here is my BERTopic version:

```
import bertopic
print(bertopic.__version__)
```
Result: `0.16.4 `

output from `topic_model.get_params()`:
```
Model params: {
'calculate_probabilities': True, 
'ctfidf_model': ClassTfidfTransformer(bm25_weighting=True, reduce_frequent_words=True), 
'embedding_model': <bertopic.backend._sentencetransformers.SentenceTransformerBackend object at 0x7f8c630c1cc0>, 'hdbscan_model': <bertopic.cluster._base.BaseCluster object at 0x7f8be1fb29b0>, 
'language': None, 
'low_memory': False, 
'min_topic_size': 10, 
'n_gram_range': (1, 1), 
'nr_topics': None, 
'representation_model': [KeyBERTInspired(), MaximalMarginalRelevance(diversity=0.5)], 
'seed_topic_list': None, 
'top_n_words': 10, 
'umap_model': UMAP(), 
'vectorizer_model': CountVectorizer(min_df=0.01,
                stop_words=['ab', 'aber', 'abermaliges', 'abermals',  'abgerufen', 'abgerufene', 'abgerufener', 'abgerufenes',  ...]), 'verbose': True, 
'zeroshot_min_similarity': 0.8, 
'zeroshot_topic_list': ['Garten', 'Versicherung', 'Hobby', 'Hilfsorganisationen', 'Mobilität', 'Auto', 'Haus', 'Immobilien', 'Energie', 'Reisen', 'Finanzen', 'Haustiere', 'Fitness']}
```

and output of `get_topic_info().head()`:

![Screenshot 2025-01-15 at 15 41 02](https://github.com/user-attachments/assets/aca8759b-d3c7-4aac-93d5-0f70a03e128f)

MISC:
# other parameters:

```
embedding_model:
  embedder: intfloat/multilingual-e5-large-instruct
  max_length_for_messages: 4500
  embedding_vector_dimension: 1024

hdbscan:
  gen_min_span_tree: True
  min_cluster_size: 60
  prediction_data: True
  min_samples: 5
  cluster_selection_epsilon: 0.35
  cluster_selection_method: 'eom'

umap:
  n_components: 10
  n_neighbors: 5
  min_dist: 0.00
  metric: cosine
```

**Reply by MaartenGr** [reply]
@alihashaam Did you try lowering `zeroshot_min_similarity`?

**Reply by alihashaam** [reply]
@MaartenGr 

Yes I did. I set it to `0` and got following results:

```
{'calculate_probabilities': True, 'ctfidf_model': ClassTfidfTransformer(bm25_weighting=True, reduce_frequent_words=True,
                      seed_words=['finanzen', 'fitness', 'hobby', 'haus',
                                  'garten', 'reisen', 'mobilität',
                                  'versicherung', 'auto', 'energie',
                                  'haustiere', 'hilfsorganisationen',
                                  'immobilien']), 
   'embedding_model': SentenceTransformer(...), 
   'hdbscan_model': HDBSCAN(), 
   'language': None, 
   'low_memory': False, 
   'min_topic_size': 10, 
   'n_gram_range': (1, 1), 
   'nr_topics': None, 
   'representation_model': [KeyBERTInspired(), MaximalMarginalRelevance(diversity=0.5)], 
   'seed_topic_list': None, 
   'top_n_words': 10, 
   'umap_model': UMAP(), 
   'vectorizer_model': CountVectorizer(min_df=0.01, stop_words=['ab', 'aber', 'abermaliges', ...]), 
   'verbose': True, 
   **'zeroshot_min_similarity': 0, 
   'zeroshot_topic_list': ['Versicherung', 'Immobilien', 'Haus', 'Finanzen', 'Auto', 'Hobby', 'Mobilität', 'Fitness', 'Garten', 'Reisen', 'Hilfsorganisationen', 'Energie', 'Haustiere']**}
```

Result of `topic_model.get_topic_info().set_index('Topic')[['Count', 'Name', 'Representation']]`:
![Screenshot 2025-01-22 at 17 23 31](https://github.com/user-attachments/assets/ff0ecf10-c44b-4046-938b-031ee8d9ac41)


So the number of topics is same as `zeroshot topic` but labelling of topics is different. 

My initial thoughts were that since we are using Embedding model "intfloat/multilingual-e5-large-instruct"  [[1]](https://huggingface.co/intfloat/multilingual-e5-large-instruct), the official documentation of the model says that "cosine similarity scores distribute around 0.7 to 1.0". 

So then trying out `.7` as `zeroshot_min_similarity` yields following:

```
{...,  
'zeroshot_min_similarity': 0.7, 
'zeroshot_topic_list': ['Energie', 'Auto', 'Haus', 'Fitness', 'Finanzen', 'Hobby', 'Garten', 'Reisen', 'Versicherung', 'Hilfsorganisationen', 'Haustiere', 'Immobilien', 'Mobilität']}
```

Topic info:
![Screenshot 2025-01-22 at 17 38 01](https://github.com/user-attachments/assets/ab6e6310-05ad-437e-a2f2-dbc1888d9d1d)

So basically with `zeroshot_min_similarity` set to  `0` or `0.7`, we have same topics. 

Following is the result of `zeroshot_min_similarity=0.75`:

[
![Screenshot 2025-01-22 at 17 44 28](https://github.com/user-attachments/assets/392a3b10-f9a2-41a9-999a-5d93039a276a)
](url)


So, if you look at provided results, we have in representations words that are related to zeroshot topic but we zeroshot topics see that in labels.

So may be there is something wrong with my process or is it about not being able to find good config of hyperparameters?

**Reply by MaartenGr** [reply]
@alihashaam Hmmm, that's odd. Could you perhaps share your full code? It may be that there is something to be found there. Also, there is a fix for something related [here](https://github.com/MaartenGr/BERTopic/pull/2267) that might help.
